// @utoo/pack@1.3.0-alpha.2/cjs/commands/build.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = build;
const pack_shared_1 = require("@utoo/pack-shared");
const child_process_1 = require("child_process");
const fs_1 = __importDefault(require("fs"));
const nanoid_1 = require("nanoid");
const path_1 = __importDefault(require("path"));
const webpackCompat_1 = require("../config/webpackCompat");
const project_1 = require("../core/project");
const HtmlPlugin_1 = require("../plugins/HtmlPlugin");
const common_1 = require("../utils/common");
const findRoot_1 = require("../utils/findRoot");
const getInitialAssets_1 = require("../utils/getInitialAssets");
const htmlEntry_1 = require("../utils/htmlEntry");
const normalize_path_1 = require("../utils/normalize-path");
const validateEntry_1 = require("../utils/validateEntry");
const xcodeProfile_1 = require("../utils/xcodeProfile");
function build(options, projectPath, rootPath) {
    const bundleOptions = (0, webpackCompat_1.resolveBundleOptions)(options, projectPath, rootPath);
    if (!rootPath) {
        // help user to find the rootDir automatically.
        rootPath = (0, findRoot_1.findRootDir)(projectPath || process.cwd());
    }
    return buildInternal(bundleOptions, projectPath, rootPath);
}
async function buildInternal(bundleOptions, projectPath, rootPath) {
    var _a, _b, _c, _d, _e, _f;
    (0, common_1.blockStdout)();
    if (process.env.XCODE_PROFILE) {
        await (0, xcodeProfile_1.xcodeProfilingReady)();
    }
    const resolvedProjectPath = projectPath || process.cwd();
    (0, htmlEntry_1.processHtmlEntry)(bundleOptions.config, resolvedProjectPath);
    (0, validateEntry_1.validateEntryPaths)(bundleOptions.config, resolvedProjectPath);
    const createProject = (0, project_1.projectFactory)();
    const project = await createProject({
        processEnv: (_a = bundleOptions.processEnv) !== null && _a !== void 0 ? _a : {},
        watch: {
            enable: false,
        },
        dev: (_b = bundleOptions.dev) !== null && _b !== void 0 ? _b : false,
        buildId: bundleOptions.buildId || (0, nanoid_1.nanoid)(),
        config: {
            ...bundleOptions.config,
            stats: Boolean(process.env.ANALYZE) ||
                bundleOptions.config.stats ||
                bundleOptions.config.entry.some((e) => !!e.html),
        },
        projectPath: (0, normalize_path_1.normalizePath)(resolvedProjectPath),
        rootPath: rootPath || projectPath || process.cwd(),
        packPath: (0, common_1.getPackPath)(),
    }, {
        persistentCaching: (_c = bundleOptions.config.persistentCaching) !== null && _c !== void 0 ? _c : false,
    });
    const entrypoints = await project.writeAllEntrypointsToDisk();
    (0, pack_shared_1.handleIssues)(entrypoints.issues);
    const htmlConfigs = [
        ...(Array.isArray(bundleOptions.config.html)
            ? bundleOptions.config.html
            : bundleOptions.config.html
                ? [bundleOptions.config.html]
                : []),
        ...bundleOptions.config.entry
            .filter((e) => !!e.html)
            .map((e) => e.html),
    ];
    if (htmlConfigs.length > 0) {
        const assets = { js: [], css: [] };
        const outputDir = ((_d = bundleOptions.config.output) === null || _d === void 0 ? void 0 : _d.path) || path_1.default.join(process.cwd(), "dist");
        if (assets.js.length === 0 && assets.css.length === 0) {
            const discovered = (0, getInitialAssets_1.getInitialAssetsFromStats)(outputDir);
            assets.js.push(...discovered.js);
            assets.css.push(...discovered.css);
        }
        const publicPath = (_e = bundleOptions.config.output) === null || _e === void 0 ? void 0 : _e.publicPath;
        for (const config of htmlConfigs) {
            const plugin = new HtmlPlugin_1.HtmlPlugin(config);
            await plugin.generate(outputDir, assets, publicPath);
        }
    }
    if (process.env.ANALYZE) {
        await analyzeBundle(((_f = bundleOptions.config.output) === null || _f === void 0 ? void 0 : _f.path) || "dist");
    }
    await project.shutdown();
    // TODO: Maybe run tasks in worker is a better way, see
    // https://github.com/vercel/next.js/blob/512d8283054407ab92b2583ecce3b253c3be7b85/packages/next/src/lib/worker.ts
}
async function analyzeBundle(outputPath) {
    const statsPath = path_1.default.join(outputPath, "stats.json");
    if (!fs_1.default.existsSync(statsPath)) {
        console.warn(`Stats file not found at ${statsPath}. Make sure to enable stats in your configuration.`);
        return;
    }
    return new Promise((resolve, reject) => {
        const analyzer = (0, child_process_1.spawn)("npx", ["webpack-bundle-analyzer", statsPath], {
            stdio: "inherit",
            shell: true,
        });
        analyzer.on("error", (error) => {
            reject(new Error(`Failed to start bundle analyzer: ${error.message}`));
        });
        analyzer.on("close", () => {
            // The analyzer process has finished, so we can resolve the promise
            // to allow the build process to exit gracefully.
            resolve();
        });
    });
}
