// @utoo/pack@1.3.0-alpha.2/cjs/commands/dev.js
"use strict";
/**
 * Dev server implementation using Hono + @hono/node-server + @hono/node-ws.
 * Keeps the same public API as dev.ts; do not remove dev.ts until this is verified.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = serve;
const node_server_1 = require("@hono/node-server");
const serve_static_1 = require("@hono/node-server/serve-static");
const node_ws_1 = require("@hono/node-ws");
const fs_1 = __importDefault(require("fs"));
const get_port_1 = __importDefault(require("get-port"));
const hono_1 = require("hono");
const https_1 = __importDefault(require("https"));
const path_1 = __importDefault(require("path"));
const webpackCompat_1 = require("../config/webpackCompat");
const hmr_1 = require("../core/hmr");
const common_1 = require("../utils/common");
const findRoot_1 = require("../utils/findRoot");
const mkcert_1 = require("../utils/mkcert");
const printServerInfo_1 = require("../utils/printServerInfo");
const xcodeProfile_1 = require("../utils/xcodeProfile");
// --- Path helpers (same logic as dev.ts, not exported) ---
function parsePath(pathStr) {
    const hashIndex = pathStr.indexOf("#");
    const queryIndex = pathStr.indexOf("?");
    const hasQuery = queryIndex > -1 && (hashIndex < 0 || queryIndex < hashIndex);
    if (hasQuery || hashIndex > -1) {
        return {
            pathname: pathStr.substring(0, hasQuery ? queryIndex : hashIndex),
            query: hasQuery
                ? pathStr.substring(queryIndex, hashIndex > -1 ? hashIndex : undefined)
                : "",
            hash: hashIndex > -1 ? pathStr.slice(hashIndex) : "",
        };
    }
    return { pathname: pathStr, query: "", hash: "" };
}
function pathHasPrefix(pathStr, prefix) {
    if (typeof pathStr !== "string") {
        return false;
    }
    const { pathname } = parsePath(pathStr);
    return pathname === prefix || pathname.startsWith(prefix + "/");
}
function removePathPrefix(pathStr, prefix) {
    if (!pathHasPrefix(pathStr, prefix)) {
        return pathStr;
    }
    const withoutPrefix = pathStr.slice(prefix.length);
    if (withoutPrefix.startsWith("/")) {
        return withoutPrefix;
    }
    return `/${withoutPrefix}`;
}
function normalizedPublicPath(publicPath) {
    const escapedPublicPath = (publicPath === null || publicPath === void 0 ? void 0 : publicPath.replace(/^\/+|\/+$/g, "")) || false;
    if (!escapedPublicPath) {
        return "";
    }
    try {
        if (URL.canParse(escapedPublicPath)) {
            const u = new URL(escapedPublicPath).toString();
            return u.endsWith("/") ? u.slice(0, -1) : u;
        }
    }
    catch (_a) { }
    return `/${escapedPublicPath}`;
}
async function resolveDevConfig(options, projectPath, rootPath, serverOptions) {
    var _a, _b, _d, _e, _f, _g;
    const cfgDevServer = (_b = (_a = options.config) === null || _a === void 0 ? void 0 : _a.devServer) !== null && _b !== void 0 ? _b : {};
    const port = typeof (serverOptions === null || serverOptions === void 0 ? void 0 : serverOptions.port) !== "undefined"
        ? serverOptions.port
        : ((_d = cfgDevServer.port) !== null && _d !== void 0 ? _d : 3000);
    const hostname = (_f = (_e = serverOptions === null || serverOptions === void 0 ? void 0 : serverOptions.hostname) !== null && _e !== void 0 ? _e : cfgDevServer.host) !== null && _f !== void 0 ? _f : "localhost";
    const useHttps = typeof (serverOptions === null || serverOptions === void 0 ? void 0 : serverOptions.https) !== "undefined"
        ? serverOptions.https
        : cfgDevServer.https;
    let selfSignedCertificate = serverOptions === null || serverOptions === void 0 ? void 0 : serverOptions.selfSignedCertificate;
    if (useHttps && !selfSignedCertificate) {
        try {
            const cert = await (0, mkcert_1.createSelfSignedCertificate)(hostname);
            if (cert)
                selfSignedCertificate = cert;
        }
        catch (_h) {
            // fall back to http
        }
    }
    const bundleOptions = {
        ...options,
        config: {
            ...options.config,
            devServer: {
                hot: true,
                ...(((_g = options.config) === null || _g === void 0 ? void 0 : _g.devServer) || {}),
            },
        },
        packPath: (0, common_1.getPackPath)(),
    };
    const projectPathResolved = projectPath || process.cwd();
    const rootPathResolved = rootPath !== null && rootPath !== void 0 ? rootPath : projectPathResolved;
    const actualPort = await (0, get_port_1.default)({ port, host: hostname });
    if (actualPort !== port) {
        console.warn(`Port ${port} is in use, using available port ${actualPort} instead.`);
    }
    const serveOptsBase = {
        port: actualPort,
        hostname,
        ...(useHttps && selfSignedCertificate
            ? {
                createServer: https_1.default.createServer,
                serverOptions: {
                    key: fs_1.default.readFileSync(selfSignedCertificate.key),
                    cert: fs_1.default.readFileSync(selfSignedCertificate.cert),
                },
            }
            : {}),
    };
    return {
        bundleOptions,
        projectPathResolved,
        rootPathResolved,
        serveOptsBase,
    };
}
// --- Entry ---
function serve(options, projectPath, rootPath, serverOptions) {
    const bundleOptions = (0, webpackCompat_1.resolveBundleOptions)(options, projectPath, rootPath);
    if (!rootPath) {
        rootPath = (0, findRoot_1.findRootDir)(projectPath || process.cwd());
    }
    return runDev(bundleOptions, projectPath, rootPath, serverOptions);
}
const HMR_PATH = "/turbopack-hmr";
async function runDev(options, projectPath, rootPath, serverOptions) {
    var _a, _b, _d, _e;
    (0, common_1.blockStdout)();
    process.title = "utoopack-dev-server";
    if (process.env.XCODE_PROFILE) {
        await (0, xcodeProfile_1.xcodeProfilingReady)();
    }
    process.env.NODE_ENV = "development";
    const { bundleOptions, projectPathResolved, rootPathResolved, serveOptsBase, } = await resolveDevConfig(options, projectPath, rootPath, serverOptions);
    const hotReloader = await (0, hmr_1.createHotReloader)(bundleOptions, projectPathResolved, rootPathResolved);
    await hotReloader.start();
    const distRoot = path_1.default.resolve(projectPathResolved, ((_b = (_a = options.config) === null || _a === void 0 ? void 0 : _a.output) === null || _b === void 0 ? void 0 : _b.path) || "./dist");
    const publicPath = (_e = (_d = options.config) === null || _d === void 0 ? void 0 : _d.output) === null || _e === void 0 ? void 0 : _e.publicPath;
    // Skip prefix stripping for "runtime" and when publicPath is absent (match dev.ts).
    const normalizedPrefix = publicPath && publicPath !== "runtime"
        ? normalizedPublicPath(publicPath)
        : "";
    const app = new hono_1.Hono();
    const { injectWebSocket, upgradeWebSocket } = (0, node_ws_1.createNodeWebSocket)({ app });
    const rewriteRequestPath = (reqPath) => {
        if (!normalizedPrefix)
            return reqPath;
        // Absolute-URL publicPath: do not rewrite (match dev.ts).
        if (normalizedPrefix.startsWith("http://") ||
            normalizedPrefix.startsWith("https://")) {
            return reqPath;
        }
        if (pathHasPrefix(reqPath, normalizedPrefix)) {
            return removePathPrefix(reqPath, normalizedPrefix);
        }
        return reqPath;
    };
    // HMR WebSocket route must be registered before "/*" so it is not handled by serveStatic
    app.get(HMR_PATH, upgradeWebSocket((_c) => ({
        onOpen(_ev, ws) {
            hotReloader.registerClient(ws);
        },
        onMessage(ev, ws) {
            try {
                const data = typeof ev.data === "string" ? ev.data : ev.data.toString();
                hotReloader.handleClientMessage(ws, data);
            }
            catch (err) {
                console.error("HMR message error", err);
            }
        },
        onClose(_ev, ws) {
            hotReloader.unregisterClient(ws);
        },
        onError(err) {
            console.error("HMR WebSocket error", err);
        },
    })));
    // GET handles HEAD automatically in Hono; serveStatic serves both
    app.get("/*", (0, serve_static_1.serveStatic)({
        root: distRoot,
        rewriteRequestPath,
    }));
    app.all("*", (c) => c.body(null, 405, { Allow: "GET, HEAD" }));
    const server = (0, node_server_1.serve)({
        ...serveOptsBase,
        fetch: app.fetch,
    });
    injectWebSocket(server);
    if ((serverOptions === null || serverOptions === void 0 ? void 0 : serverOptions.logServerInfo) !== false) {
        const scheme = serveOptsBase.serverOptions ? "https" : "http";
        const displayHost = serveOptsBase.hostname === "0.0.0.0"
            ? "localhost"
            : serveOptsBase.hostname;
        (0, printServerInfo_1.printServerInfo)(scheme, displayHost, serveOptsBase.port);
    }
    const cleanup = () => {
        hotReloader.close();
        // We always create HTTP/1.1 server (http or https), so closeAllConnections exists; Hono's
        // ServerType union includes HTTP/2, so TS does not narrow. Use runtime check to satisfy types.
        if ("closeAllConnections" in server &&
            typeof server.closeAllConnections === "function") {
            server.closeAllConnections();
        }
        server.close((err) => {
            if (err) {
                console.error(err);
                process.exit(1);
            }
            process.exit(0);
        });
    };
    const exception = (err) => {
        console.error(err);
    };
    process.on("SIGINT", cleanup);
    process.on("SIGTERM", cleanup);
    process.on("rejectionHandled", () => { });
    process.on("uncaughtException", exception);
    process.on("unhandledRejection", exception);
}
