// @utoo/pack-shared@1.3.0-alpha.2/cjs/webpackCompat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compatOptionsFromWebpack = compatOptionsFromWebpack;
function compatOptionsFromWebpack(webpackConfig) {
    const { entry, mode, module, resolve, externals, output, target, devtool, optimization, plugins, stats, } = webpackConfig;
    const html = compatFromWebpackPlugin(plugins, compatHtml);
    const copy = compatFromWebpackPlugin(plugins, compatCopy);
    const outputCompat = compatOutput(output);
    if (outputCompat && !outputCompat.cssFilename) {
        outputCompat.cssFilename = compatFromWebpackPlugin(plugins, compatMiniCssExtract);
    }
    if (outputCompat && copy) {
        outputCompat.copy = copy;
    }
    return {
        config: {
            entry: compatEntry(entry, html),
            mode: compatMode(mode),
            module: compatModule(module),
            resolve: compatResolve(resolve),
            externals: compatExternals(externals),
            output: outputCompat,
            target: compatTarget(target),
            sourceMaps: compatSourceMaps(devtool),
            optimization: compatOptimization(optimization),
            define: compatFromWebpackPlugin(plugins, compatDefine),
            provider: compatFromWebpackPlugin(plugins, compatProvider),
            stats: compatStats(stats),
            devServer: compatDevServer(webpackConfig.devServer),
        },
        buildId: webpackConfig.name,
    };
}
function compatMode(webpackMode) {
    return webpackMode
        ? webpackMode === "none"
            ? "production"
            : webpackMode
        : "production";
}
function compatEntry(webpackEntry, html) {
    if (!webpackEntry) {
        return undefined;
    }
    const entry = [];
    switch (typeof webpackEntry) {
        case "string":
            entry.push({ import: webpackEntry, html });
            break;
        case "object":
            if (Array.isArray(webpackEntry)) {
                webpackEntry.forEach((e) => entry.push({
                    import: e,
                    html,
                }));
            }
            else {
                Object.entries(webpackEntry).forEach(([k, v]) => {
                    var _a;
                    switch (typeof v) {
                        case "string":
                            entry.push({ name: k, import: v, html });
                            break;
                        case "object":
                            if (!Array.isArray(v)) {
                                switch (typeof v.import) {
                                    case "string":
                                        entry.push({
                                            name: k,
                                            import: v.import,
                                            html,
                                            library: ((_a = v.library) === null || _a === void 0 ? void 0 : _a.type) === "umd"
                                                ? {
                                                    name: typeof v.library.name === "string"
                                                        ? v.library.name
                                                        : undefined,
                                                    export: typeof v.library.export === "string"
                                                        ? [v.library.export]
                                                        : v.library.export,
                                                }
                                                : undefined,
                                        });
                                        break;
                                    default:
                                        break;
                                }
                            }
                            else {
                                if (v.length === 0) {
                                    throw "entry value items is empty";
                                }
                                else if (v.length === 1) {
                                    entry.push({ name: k, import: v[0], html });
                                }
                                else {
                                    throw "multi entry items for one entry not supported yet";
                                }
                            }
                            break;
                        default:
                            throw "non string and non object entry path not supported yet";
                    }
                });
            }
            break;
        case "function":
            throw new Error("functional entry not supported yet");
        default:
            throw new Error("entry config not compatible now");
    }
    return entry;
}
function compatFromWebpackPlugin(webpackPlugins, picker) {
    const plugin = webpackPlugins === null || webpackPlugins === void 0 ? void 0 : webpackPlugins.find((p) => p && typeof p === "object" && p.constructor.name === picker.pluginName);
    return picker(plugin);
}
compatHtml.pluginName = "HtmlWebpackPlugin";
function compatHtml(maybeWebpackPluginInstance) {
    if (!maybeWebpackPluginInstance) {
        return undefined;
    }
    return (maybeWebpackPluginInstance.userOptions ||
        maybeWebpackPluginInstance.options);
}
compatMiniCssExtract.pluginName = "MiniCssExtractPlugin";
function compatMiniCssExtract(maybeWebpackPluginInstance) {
    var _a;
    return (_a = maybeWebpackPluginInstance === null || maybeWebpackPluginInstance === void 0 ? void 0 : maybeWebpackPluginInstance.options) === null || _a === void 0 ? void 0 : _a.filename;
}
compatCopy.pluginName = "CopyPlugin";
function compatCopy(maybeWebpackPluginInstance) {
    const patterns = maybeWebpackPluginInstance === null || maybeWebpackPluginInstance === void 0 ? void 0 : maybeWebpackPluginInstance.patterns;
    if (!Array.isArray(patterns))
        return undefined;
    return patterns.map((pattern) => {
        if (typeof pattern === "string")
            return pattern;
        return {
            from: pattern.from,
            to: pattern.to,
        };
    });
}
compatDefine.pluginName = "DefinePlugin";
function compatDefine(maybeWebpackPluginInstance) {
    const definitions = maybeWebpackPluginInstance === null || maybeWebpackPluginInstance === void 0 ? void 0 : maybeWebpackPluginInstance.definitions;
    if (!definitions || typeof definitions !== "object") {
        return definitions;
    }
    const processedDefinitions = {};
    for (const [key, value] of Object.entries(definitions)) {
        if (typeof value === "object" && value !== null) {
            processedDefinitions[key] = JSON.stringify(value);
        }
        else {
            processedDefinitions[key] = value;
        }
    }
    return processedDefinitions;
}
compatProvider.pluginName = "ProvidePlugin";
function compatProvider(maybeWebpackPluginInstance) {
    const definitions = maybeWebpackPluginInstance === null || maybeWebpackPluginInstance === void 0 ? void 0 : maybeWebpackPluginInstance.definitions;
    if (!definitions || typeof definitions !== "object") {
        return undefined;
    }
    const provider = {};
    for (const [key, value] of Object.entries(definitions)) {
        if (typeof value === "string") {
            // Simple module import: { $: 'jquery' }
            provider[key] = value;
        }
        else if (Array.isArray(value)) {
            if (value.length === 1) {
                // e.g. { process: ['process/browser'] } for default import
                provider[key] = value[0];
            }
            else if (value.length >= 2) {
                // Named export import: { Buffer: ['buffer', 'Buffer'] }
                provider[key] = [value[0], value[1]];
            }
        }
    }
    return Object.keys(provider).length > 0 ? provider : undefined;
}
function compatExternals(webpackExternals) {
    if (!webpackExternals) {
        return undefined;
    }
    let externals = {};
    switch (typeof webpackExternals) {
        case "string": {
            // Single string external: "lodash" -> { "lodash": "lodash" }
            externals[webpackExternals] = webpackExternals;
            break;
        }
        case "object": {
            if (Array.isArray(webpackExternals)) {
                // Array of externals: ["lodash", "react"] -> { "lodash": "lodash", "react": "react" }
                externals = webpackExternals.reduce((acc, external) => {
                    if (typeof external === "string") {
                        acc[external] = external;
                    }
                    else if (typeof external === "object" && external !== null) {
                        Object.assign(acc, compatExternals(external));
                    }
                    return acc;
                }, {});
            }
            else if (webpackExternals instanceof RegExp) {
                throw new Error("regex external not supported yet");
            }
            else {
                if ("byLayer" in webpackExternals) {
                    throw new Error("by layer external item not supported yet");
                }
                Object.entries(webpackExternals).forEach(([key, value]) => {
                    if (typeof value === "string") {
                        // Check if it's a script type with shorthand syntax: "global@https://example.com/script.js"
                        if (value.includes("@") &&
                            (value.startsWith("script ") || value.includes("://"))) {
                            const match = value.match(/^(?:script\s+)?(.+?)@(.+)$/);
                            if (match) {
                                const [, globalName, scriptUrl] = match;
                                // Use utoopack string format: "script globalName@url"
                                externals[key] = `script ${globalName}@${scriptUrl}`;
                            }
                            else {
                                externals[key] = value;
                            }
                        }
                        else {
                            // Simple string mapping: { "react": "React" }
                            externals[key] = value;
                        }
                    }
                    else if (Array.isArray(value)) {
                        // Array format handling
                        if (value.length >= 2) {
                            const [first, second] = value;
                            // Check if it's a script type array: ["https://example.com/script.js", "GlobalName"]
                            if (typeof first === "string" &&
                                first.includes("://") &&
                                typeof second === "string") {
                                // Use utoopack object format for script
                                externals[key] = {
                                    root: second,
                                    type: "script",
                                    script: first,
                                };
                            }
                            else if (typeof first === "string" &&
                                typeof second === "string") {
                                // Handle type prefix formats
                                if (first.startsWith("commonjs")) {
                                    externals[key] = `commonjs ${second}`;
                                }
                                else if (first === "module") {
                                    externals[key] = `esm ${second}`;
                                }
                                else if (first === "var" ||
                                    first === "global" ||
                                    first === "window") {
                                    externals[key] = second;
                                }
                                else if (first === "script") {
                                    // Script type without URL in array format - treat as regular script prefix
                                    externals[key] = `script ${second}`;
                                }
                                else {
                                    externals[key] = `${first} ${second}`;
                                }
                            }
                            else {
                                externals[key] = value[0] || key;
                            }
                        }
                        else {
                            externals[key] = value[0] || key;
                        }
                    }
                    else if (typeof value === "object" && value !== null) {
                        // Object format: handle complex configurations
                        const val = value;
                        if ("root" in val || "commonjs" in val || "amd" in val) {
                            // Standard webpack externals object format
                            if (val.commonjs) {
                                externals[key] = `commonjs ${val.commonjs}`;
                            }
                            else if (val.root) {
                                externals[key] = val.root;
                            }
                            else if (val.amd) {
                                externals[key] = val.amd;
                            }
                            else {
                                externals[key] = key;
                            }
                        }
                        else {
                            // Treat as utoopack specific configuration (might already be in correct format)
                            externals[key] = value;
                        }
                    }
                    else {
                        // Fallback to key name
                        externals[key] = key;
                    }
                });
            }
            break;
        }
        case "function": {
            throw new Error("functional external not supported yet");
        }
        default:
            break;
    }
    return externals;
}
function compatModule(webpackModule) {
    if (!Array.isArray(webpackModule === null || webpackModule === void 0 ? void 0 : webpackModule.rules)) {
        return;
    }
    const moduleRules = {
        rules: webpackModule.rules.reduce((acc, cur) => {
            var _a;
            if (typeof cur === "object" && cur) {
                // Normalize condition from test, include, exclude
                let ruleCondition = undefined;
                if (cur.test || cur.include || cur.exclude) {
                    const conditions = [];
                    if (cur.test)
                        conditions.push({ path: cur.test });
                    if (cur.include)
                        conditions.push({ path: cur.include });
                    if (cur.exclude)
                        conditions.push({ not: { path: cur.exclude } });
                    ruleCondition =
                        conditions.length === 1 ? conditions[0] : { all: conditions };
                }
                // Determine the glob keys for the rules record
                // utoo usually uses extensions as keys like "*.svg"
                const extensions = (_a = cur.test) === null || _a === void 0 ? void 0 : _a.toString().match(/\.(\w+)/g);
                const globKeys = extensions
                    ? extensions.map((ext) => `*${ext}`)
                    : ["*"];
                // Handle loaders
                const loaders = typeof cur.use === "string"
                    ? [{ loader: cur.use, options: {} }]
                    : Array.isArray(cur.use)
                        ? cur.use.map((use) => typeof use === "string"
                            ? { loader: use, options: {} }
                            : {
                                loader: use.loader,
                                options: use.options || {},
                            })
                        : cur.loader
                            ? [
                                {
                                    loader: cur.loader,
                                    options: cur.options || {},
                                },
                            ]
                            : [];
                const ruleItem = {
                    loaders,
                    condition: ruleCondition,
                    as: cur.type === "asset" ||
                        cur.type === "asset/resource" ||
                        cur.type === "asset/inline" ||
                        cur.type === "asset/source" ||
                        (loaders.length > 0 &&
                            !globKeys.some((k) => k === "*.js" || k === "*.ts" || k === "*.tsx"))
                        ? "*.js"
                        : undefined,
                };
                for (const key of globKeys) {
                    const existing = acc[key];
                    if (!existing) {
                        acc[key] = ruleItem;
                    }
                    else {
                        // If already exists, turn into a collection array
                        if (Array.isArray(existing)) {
                            existing.push(ruleItem);
                        }
                        else {
                            acc[key] = [existing, ruleItem];
                        }
                    }
                }
            }
            return acc;
        }, {}),
    };
    return moduleRules;
}
function compatResolve(webpackResolve) {
    if (!webpackResolve) {
        return;
    }
    const { alias, extensions } = webpackResolve;
    return {
        alias: alias
            ? Array.isArray(alias)
                ? alias.reduce((acc, cur) => Object.assign(acc, { [cur.name]: cur.alias }), {})
                : Object.entries(alias).reduce((acc, [k, v]) => {
                    if (typeof v === "string") {
                        // Handle alias keys ending with $ by removing the $
                        const aliasKey = k.endsWith("$") ? k.slice(0, -1) : k;
                        acc[aliasKey] = v;
                    }
                    else {
                        throw new Error("non string alias value not supported yet");
                    }
                    return acc;
                }, {})
            : undefined,
        extensions,
    };
}
function normalizeWebpackHash(filename) {
    if (!filename)
        return filename;
    // Replace [hash] and [chunkhash] with [contenthash] as utoo only supports contenthash
    // Also remove [ext] as it is automatically appended in utoopack
    return filename
        .replace(/\[(?:hash|chunkhash)(?::(\d+))?\]/g, "[contenthash:$1]")
        .replace(/\[ext\]/g, "");
}
function compatOutput(webpackOutput) {
    for (const field of [
        "filename",
        "chunkFilename",
        "publicPath",
        "cssFilename",
        "cssChunkFilename",
        "assetModuleFilename",
    ]) {
        if ((webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput[field]) && typeof webpackOutput[field] !== "string") {
            throw new Error(`non string output ${field} not supported yet`);
        }
    }
    return {
        path: webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.path,
        filename: normalizeWebpackHash(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.filename),
        chunkFilename: normalizeWebpackHash(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.chunkFilename),
        cssFilename: normalizeWebpackHash(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.cssFilename),
        cssChunkFilename: normalizeWebpackHash(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.cssChunkFilename),
        assetModuleFilename: normalizeWebpackHash(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.assetModuleFilename),
        clean: !!(webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.clean),
        publicPath: webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.publicPath,
        chunkLoadingGlobal: webpackOutput === null || webpackOutput === void 0 ? void 0 : webpackOutput.chunkLoadingGlobal,
    };
}
function compatTarget(webpackTarget) {
    return webpackTarget
        ? Array.isArray(webpackTarget)
            ? webpackTarget.join(" ")
            : webpackTarget
        : undefined;
}
function compatSourceMaps(webpackSourceMaps) {
    return !!webpackSourceMaps;
}
function compatOptimization(webpackOptimization) {
    if (!webpackOptimization) {
        return;
    }
    const { moduleIds, minimize, concatenateModules, usedExports } = webpackOptimization;
    const enableWebpackUsedExports = usedExports !== false;
    return {
        moduleIds: moduleIds === "named"
            ? "named"
            : moduleIds === "deterministic"
                ? "deterministic"
                : undefined,
        noMangling: webpackOptimization.mangleExports === false,
        minify: minimize,
        concatenateModules,
        treeShaking: false,
        removeUnusedExports: enableWebpackUsedExports,
        removeUnusedImports: enableWebpackUsedExports,
    };
}
function compatStats(webpackStats) {
    return !!webpackStats;
}
function compatDevServer(devServer) {
    if (!devServer)
        return undefined;
    const result = {};
    if (typeof devServer.hot !== "undefined")
        result.hot = !!devServer.hot;
    if (typeof devServer.port !== "undefined") {
        const p = Number(devServer.port);
        if (!Number.isNaN(p))
            result.port = p;
    }
    if (typeof devServer.host !== "undefined")
        result.host = devServer.host;
    if (typeof devServer.https !== "undefined")
        result.https = !!devServer.https;
    return result;
}
